File: Third Semester - Advanced Programming Methods/SocialNetwork/java/org/example/socialnetwork/repository/inmemory/InMemoryRepository.java
package org.example.socialnetwork.repository.inmemory;



import org.example.socialnetwork.domain.Entity;
import org.example.socialnetwork.domain.validators.Validator;
import org.example.socialnetwork.repository.Repository;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * A generic in memory repository implementation for managing entities.
 * @param <ID> the type of the entity ID
 * @param <E> type of entities saved in repository
 */
public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {
    private final Validator<E> validator;
    protected Map<ID, E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        this.entities = new HashMap<>();
    }

    @Override
    public Optional<E> findOne(ID id) {
        if (id == null) {
            throw new IllegalArgumentException("id cannot be null");
        }
        return Optional.ofNullable(entities.get(id));
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public Optional<E> save(E entity) {
        if (entity == null) {
            throw new IllegalArgumentException("id must not be null");
        }
        validator.validate(entity);
        return Optional.ofNullable(entities.putIfAbsent(entity.getId(), entity));
    }
    @Override
    public Optional<E> delete(ID id) {
        if (id == null) {
            throw new IllegalArgumentException("id must not be null");
        }
        return Optional.ofNullable(entities.remove(id));
    }

    @Override
    public Optional<E> update(E entity) {
        if(entity == null) {
            throw new IllegalArgumentException("entity cannot be null");
        }
        validator.validate(entity);
        if(entities.containsKey(entity.getId())) {
            entities.put(entity.getId(), entity);
            return Optional.empty();
        }
        return Optional.of(entity);
    }
}

File: Third Semester - Advanced Programming Methods/Lab11/repository/inmemory/InMemoryRepository.cs
using lab11.domain;

namespace lab11.repository.inmemory;

public class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
{
    protected Dictionary<ID, E?> Entities = new Dictionary<ID, E?>();

    public virtual E? Add(E? entity)
    {
        ArgumentNullException.ThrowIfNull(entity);
        return Entities.TryAdd(entity.Id, entity) ? entity : default;
    }

    public virtual E? Delete(ID id)
    {
        ArgumentNullException.ThrowIfNull(id);
        Entities.Remove(id, out E? entity);
        return entity;
    }

    public virtual E? Update(E? entity)
    {
        ArgumentNullException.ThrowIfNull(entity);
        if (!Entities.TryGetValue(entity.Id, out var update))
        {
            return default;
        }
        return update;
    }

    public virtual E? Get(ID id)
    {
       ArgumentNullException.ThrowIfNull(id);
       return Entities.GetValueOrDefault(id);
    }

    public virtual IEnumerable<E> GetAll()
    {
        return Entities.Values;
    }
}